File: pyspider/examples/ecommerce_crawler.py
from __future__ import annotations

import sys
from dataclasses import dataclass
from pathlib import Path
from typing import Any

from pyspider.spider.spider import CrawlerProcess, FeedExporter, Spider

from ecommerce_browser_spider import EcommerceBrowserSpider
from ecommerce_catalog_spider import EcommerceCatalogSpider
from ecommerce_detail_spider import EcommerceDetailSpider
from ecommerce_review_spider import EcommerceReviewSpider
from ecommerce_site_profile import DEFAULT_SITE_FAMILY


@dataclass
class EcommerceCrawlerResult:
    mode: str
    item_count: int
    output_path: str
    payload: dict[str, Any] | None = None


class EcommerceCrawler:
    """Unified ecommerce crawler class for static and browser-backed modes."""

    def __init__(self, site_family: str = DEFAULT_SITE_FAMILY, output_dir: str = "artifacts/exports"):
        self.site_family = site_family
        self.output_dir = output_dir

    def build_spider(self, mode: str) -> tuple[Spider, str]:
        normalized = self.normalize_mode(mode)
        if normalized == "detail":
            return EcommerceDetailSpider(site_family=self.site_family), normalized
        if normalized == "review":
            return EcommerceReviewSpider(site_family=self.site_family), normalized
        return EcommerceCatalogSpider(site_family=self.site_family), "catalog"

    def run(self, mode: str = "catalog") -> EcommerceCrawlerResult:
        if mode in {"browser", "selenium", "playwright"}:
            return self.run_browser("catalog")

        spider, normalized_mode = self.build_spider(mode)
        items = CrawlerProcess(spider).start()
        output = Path(self.output_dir)
        output.mkdir(parents=True, exist_ok=True)
        output_path = str(output / f"pyspider-{self.site_family}-{normalized_mode}.json")
        exporter = FeedExporter.json(output_path)
        exporter.export_items(items)
        exported = exporter.close()
        return EcommerceCrawlerResult(mode=normalized_mode, item_count=len(items), output_path=exported)

    def run_browser(self, mode: str = "catalog", output_dir: str = "artifacts/browser") -> EcommerceCrawlerResult:
        normalized_mode = self.normalize_mode(mode)
        payload = EcommerceBrowserSpider(site_family=self.site_family, output_dir=output_dir).crawl(normalized_mode)
        json_path = payload.get("artifacts", {}).get("json", "")
        product_count = len(payload.get("json_ld_products", [])) + len(payload.get("bootstrap_products", []))
        return EcommerceCrawlerResult(
            mode=normalized_mode,
            item_count=product_count,
            output_path=json_path,
            payload=payload,
        )

    @staticmethod
    def normalize_mode(mode: str) -> str:
        return mode if mode in {"catalog", "detail", "review"} else "catalog"


if __name__ == "__main__":
    selected_mode = sys.argv[1] if len(sys.argv) > 1 else "catalog"
    selected_family = sys.argv[2] if len(sys.argv) > 2 else DEFAULT_SITE_FAMILY
    crawler = EcommerceCrawler(site_family=selected_family)
    if selected_mode in {"browser", "selenium", "playwright"}:
        browser_mode = sys.argv[3] if len(sys.argv) > 3 else "catalog"
        result = crawler.run_browser(browser_mode)
    else:
        result = crawler.run(selected_mode)
    print(f"exported {result.item_count} items to {result.output_path}")

File: javaspider/src/main/java/com/javaspider/examples/ecommerce/EcommerceCrawler.java
package com.javaspider.examples.ecommerce;

import com.javaspider.scrapy.CrawlerProcess;
import com.javaspider.scrapy.Spider;
import com.javaspider.scrapy.feed.FeedExporter;
import com.javaspider.scrapy.item.Item;

import java.nio.file.Files;
import java.nio.file.Path;
import java.util.List;
import java.util.Map;

public final class EcommerceCrawler {
    private final String siteFamily;
    private final Path outputDir;

    public EcommerceCrawler() {
        this(EcommerceSiteProfiles.DEFAULT_SITE_FAMILY, Path.of("artifacts", "exports"));
    }

    public EcommerceCrawler(String siteFamily) {
        this(siteFamily, Path.of("artifacts", "exports"));
    }

    public EcommerceCrawler(String siteFamily, Path outputDir) {
        this.siteFamily = siteFamily;
        this.outputDir = outputDir;
    }

    public Spider buildSpider(String mode) {
        return switch (normalizeMode(mode)) {
            case "detail" -> new EcommerceDetailSpider(siteFamily);
            case "review" -> new EcommerceReviewSpider(siteFamily);
            default -> new EcommerceCatalogSpider(siteFamily);
        };
    }

    public CrawlResult run(String mode) throws Exception {
        String normalizedMode = normalizeMode(mode);
        List<Item> items = new CrawlerProcess(buildSpider(normalizedMode)).crawl();
        Files.createDirectories(outputDir);
        Path outputPath = outputDir.resolve("javaspider-" + siteFamily + "-" + normalizedMode + ".json");
        try (FeedExporter exporter = FeedExporter.json(outputPath.toString())) {
            exporter.exportItems(items);
        }
        return new CrawlResult(normalizedMode, items.size(), outputPath.toString(), Map.of());
    }

    public CrawlResult runBrowser(String mode) throws Exception {
        String normalizedMode = normalizeMode(mode);
        Map<String, Object> payload = EcommerceSeleniumCrawler.capture(
            siteFamily,
            normalizedMode,
            Path.of("artifacts", "browser")
        );
        @SuppressWarnings("unchecked")
        Map<String, String> artifacts = payload.get("artifacts") instanceof Map<?, ?> raw
            ? (Map<String, String>) raw
            : Map.of();
        int productCount = listSize(payload.get("json_ld_products")) + listSize(payload.get("bootstrap_products"));
        return new CrawlResult(
            normalizedMode,
            productCount,
            artifacts.getOrDefault("json", ""),
            payload
        );
    }

    private static int listSize(Object value) {
        return value instanceof List<?> list ? list.size() : 0;
    }

    private static String normalizeMode(String mode) {
        return switch (mode == null ? "" : mode) {
            case "detail" -> "detail";
            case "review" -> "review";
            default -> "catalog";
        };
    }

    public record CrawlResult(String mode, int itemCount, String outputPath, Map<String, Object> payload) {
    }

    public static void main(String[] args) throws Exception {
        String mode = args.length > 0 ? args[0] : "catalog";
        String siteFamily = args.length > 1 ? args[1] : EcommerceSiteProfiles.DEFAULT_SITE_FAMILY;
        EcommerceCrawler crawler = new EcommerceCrawler(siteFamily);
        CrawlResult result;
        if ("browser".equals(mode) || "selenium".equals(mode)) {
            String browserMode = args.length > 2 ? args[2] : "catalog";
            result = crawler.runBrowser(browserMode);
        } else {
            result = crawler.run(mode);
        }
        System.out.println("exported " + result.itemCount() + " items to " + result.outputPath());
    }
}
